aten/src/ATen/FunctionalStorageImpl.cpp

```diff
@@ -0,0 +1,110 @@
+#include <ATen/FunctionalStorageImpl.h>
+
+#include <ATen/FunctionalTensorWrapper.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+#include <vector>
+
+namespace at {
+namespace functionalization {
+
+Alias::Alias(const at::Tensor& base) {
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(base));
+  base_ = base;
+}
+
+const at::Tensor& Alias::base() const {
+  return base_;
+}
+
+// metas is taken by value on purpose - we want to copy the vector.
+void Alias::add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas) {
+  updates_.push_back({updated_val, metas});
+  generation_++;
+}
+
+// Note [Functionalization: Alias Removal Part 2]
+// See Note [Functionalization: Alias Removal] for more details.
+// This function applies a single update from one of the views to the Alias object.
+// We start out with <original_base> and <mutated_view>, and our goal is to end up with <mutated_base>.
+void Alias::apply_update(const Update& update) {
+  at::AutoDispatchBelowFunctionalize guard;
+  at::Tensor t = update.new_val;
+  TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(t));
+  std::vector<at::Tensor> tmp_values({base_});
+  // First, we replay each view on <original_base>.
+  // We only actually need this for ops like select/slice/diagonal, which creates a view that's a subset of the original tensor.
+  // e.g.:
+  // a = torch.ones(...)
+  // b = a[0]
+  // c = a.permute(...)
+  // d = c[0]
+  // d.add_(1)
```

Comment:
I would have appreciated a closer correspondence between the pseudocode here and `update.view_metas`, and in general the algorithm below. This would also be easier if the example here didn't involve the same indexing operation so you can't easily distinguish them. Here is what I'd have liked to see:

```
base = ...
a = base.view1()
b = a.view2()
c = b.view3()
c.add_(3)

Then we have:

update.view_metas = [view1, view2, view3]
tmp_values = [base, a, b]  # NB: c is not necessary

And we end invoking these inverses:

t = update.new_val
t = view3_inverse(b, t, 0)  # 0 is output index, these are all single output views so it's 0
t = view2_inverse(a, t, 0)
t = view1_inverse(base, t, 0)

Inverses do something like this:

t = [ TTT ]
b = [ BBB BBB BBB ]
view3_inverse(b, t, 0) = [ TTT BBB BBB ]
```

Reply:
You're right, I don't have an documentation that clearly describes the level-down mapping from a functionalized program to the Alias/ViewMeta API calls. I'll add a flavor of what you wrote here.